### ISSUE TITLE
Error when using #[wasm_bindgen] on struct named Response

### ISSUE THREAD

**Issue by DrGabble** [rustwasm/wasm-bindgen#2470]
With the wasm rust code:
```rust
use wasm_bindgen::prelude::*;

#[wasm_bindgen]
pub struct Response {}

#[wasm_bindgen]
pub fn foo() -> Response {
    Response {}
}
```

And the html / js:
```html
<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8" />
  <title>Foo</title>
  <script src="./pkg/wasm.js"></script>
</head>

<body>

  <script type="module">
    import { default as wasm, foo } from "./pkg/wasm.js";
    wasm().then((module) => {
      var response = foo();
    });

  </script>

</body>

</html>
```

I'm getting the js error during the init process:

```
TypeError: first argument must be a WebAssembly.Module, ArrayBuffer or typed array object
```

![Screenshot from 2021-02-27 17-18-27](https://user-images.githubusercontent.com/15156457/109394756-81e33e80-7920-11eb-94c6-388a385daadd.png)

Uniquely, doesn't seem to be an issue if I change the struct name to anything else e.g. Foo. Presumably this export is having a name collision with something js side?

**Comment by alexcrichton** [comment]
I suspect this is having a collision, yes, but I don't exactly know how. Would be worthwhile to investigate! I unfortunately don't have time to do so right now.

**Comment by DrGabble** [comment]
Yeah sorry, likewise me too or I would have dug further! No worries, as I said the workaround is easy, just thought I'd let you know for posterity etc

**Comment by tari** [comment]
I bumped into this too. It looks like the exported `Response` class causes `load(module, imports)` to misbehave. In the generated js, simplified somewhat:

```javascript
class Response {
    /* bindgen'd Response class */
}
__exports.Response = Response;

async function load(module, imports) {
    if (typeof Response === 'function' && module instanceof Response) {
        if (typeof WebAssembly.instantiateStreaming === 'function') {
            // try to instantiate streaming
        }
        const bytes = await module.arrayBuffer();
        return await WebAssembly.instantiate(bytes, imports);
    } else {
        // Assumed to have been passed an existing module or buffer
        WebAssembly.instantiate(module, imports);
    }
}
```

In typical use `init('path.wasm')` will `fetch` the provided URL and pass that to `load` which checks whether the provided module is a streaming response in order to use `instantiateStreaming` or fall back to loading the content into a buffer. The user-defined `Response` shadows the fetch `Response` type so it instead falls through to the `else` branch which assumes the provided `module` is already suitable for direct instantiation.

**Comment by tari** [comment]
It seems like the parts of the javascript shims that do type checking using global types like `Response` and `WebAssembly` should qualify them through [`globalThis`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis) to avoid any potential shadowing from user types.
